src/core/engine/ffmpeg/ffmpegreplaygain.cpp

```diff
@@ -0,0 +1,383 @@
+
+    if(rc = avcodec_open2(decoderCtx, decoder, nullptr); rc < 0) {
+        Utils::printError(rc);
+        return false;
+    }
+
+    m_format = Utils::audioFormatFromCodec(stream.avStream()->codecpar);
+    filter   = initialiseRGFilter(m_format);
+
+    return true;
+}
+
+ReplayGainResult FFmpegReplayGainPrivate::handleTrack(bool inAlbum) const
+{
+    int rc{0};
+    const FramePtr framePtr{av_frame_alloc()};
+    const PacketPtr packetPtr{av_packet_alloc()};
+    auto* frame  = framePtr.get();
+    auto* packet = packetPtr.get();
+
+    while(rc == 0) {
+        rc = av_read_frame(m_formatCtx.get(), packet);
+        if(rc < 0) {
+            break;
+        }
+
+        if(packet->stream_index != m_streamIndex) {
+            av_packet_unref(packet);
+            continue;
+        }
+
+        rc = avcodec_send_packet(m_decoderCtx.get(), packet);
+        if(rc < 0) {
+            Utils::printError(rc);
+            break;
+        }
+
+        while(rc >= 0) {
+            rc = avcodec_receive_frame(m_decoderCtx.get(), frame);
+            if(rc == AVERROR_EOF || rc == AVERROR(EAGAIN)) {
+                rc = 0;
+                break;
+            }
+            if(rc < 0) {
+                Utils::printError(rc);
+                break;
+            }
+            rc = av_buffersrc_add_frame_flags(m_trackFilter.filterContext, frame, FrameFlags);
+            if(rc < 0) {
+                Utils::printError(rc);
+                break;
+            }
+            if(inAlbum) {
+                rc = av_buffersrc_add_frame_flags(m_albumFilter.filterContext, frame, FrameFlags);
+                if(rc < 0) {
+                    Utils::printError(rc);
+                    break;
+                }
+            }
+            av_frame_unref(frame);
+        }
+
+        av_packet_unref(packet);
+    }
+
+    if(!finishFilter(m_trackFilter.filterContext)) {
+        return {};
+    }
+
+    return extractRGValues(m_trackFilter.filterGraph.get());
+}
+
+void FFmpegReplayGainPrivate::handleAlbum(const TrackList& album)
+{
+    // FIXME
+    if(!setupTrack(album[0], m_albumFilter)) {
+        return;
+    }
+
+    if(!m_albumFilter.filterContext || !m_albumFilter.filterGraph) {
+        return;
+    }
+
+    std::vector<ReplayGainResult> trackResults;
+
+    for(auto track : album) {
+        if(setupTrack(track, m_trackFilter)) {
+            trackResults.emplace_back(handleTrack(true));
+        }
+    }
+
+    if(!finishFilter(m_albumFilter.filterContext)) {
+        return;
+    }
+
+    const auto albumResult = extractRGValues(m_albumFilter.filterGraph.get());
+
+    for(size_t i = 0; auto track : album) {
+        track.setRGTrackGain(static_cast<float>(trackResults[i].gain));
+        track.setRGTrackPeak(static_cast<float>(trackResults[i].peak));
+        track.setRGAlbumPeak(static_cast<float>(albumResult.peak));
+        track.setRGAlbumGain(static_cast<float>(albumResult.gain));
+        m_library->writeTrackMetadata({track});
```

Comment:
`writeTrackMetadata` is quite expensive, so it should be moved outside of the loop and a single call used for the entire album.